pkg/syncer/metadatasyncer.go

```diff
@@ -517,33 +546,46 @@ func updatePodMetadata(ctx context.Context, pod *v1.Pod, metadataSyncer *metadat
 func csiPVCUpdated(ctx context.Context, pvc *v1.PersistentVolumeClaim, pv *v1.PersistentVolume, metadataSyncer *metadataSyncInformer) {
 	log := logger.GetLogger(ctx)
 	volumeFound := false
-	// Following wait poll is required to avoid race condition between pvcUpdated and pvUpdated
-	// This helps avoid race condition between pvUpdated and pvcUpdated handlers when static PV and PVC is created almost
-	// at the same time using single YAML file.
-	err := wait.Poll(5*time.Second, time.Minute, func() (bool, error) {
-		queryFilter := cnstypes.CnsQueryFilter{
-			VolumeIds: []cnstypes.CnsVolumeId{{Id: pv.Spec.CSI.VolumeHandle}},
+	var volumeHandle string
+	var err error
+	if metadataSyncer.configInfo.Cfg.FeatureStates.CSIMigration && pv.Spec.VsphereVolume != nil {
+		volumeHandle, err = GetVolumeID(ctx, pv)
+		if err != nil {
+			log.Errorf("PVCDeleted: failed to get volume id for volume name: %q with err=%v", pv.Name, err)
+			return
 		}
-		queryResult, err := metadataSyncer.volumeManager.QueryVolume(ctx, queryFilter)
+	} else {
+		volumeHandle = pv.Spec.CSI.VolumeHandle
+		// Following wait poll is required to avoid race condition between pvcUpdated and pvUpdated
+		// This helps avoid race condition between pvUpdated and pvcUpdated handlers when static PV and PVC is created almost
+		// at the same time using single YAML file.
+		err := wait.Poll(5*time.Second, time.Minute, func() (bool, error) {
+			queryFilter := cnstypes.CnsQueryFilter{
+				VolumeIds: []cnstypes.CnsVolumeId{{Id: volumeHandle}},
+			}
+			queryResult, err := metadataSyncer.volumeManager.QueryVolume(ctx, queryFilter)
```

Comment:
I think here you want to wait in pvcUpdated until pvUpdated is done with registering volume as CNS volume.
Instead of query volume, Can we rely on a local cache to avoid this race?

Reply:
Query volume is only called for CSI volumes. For migrated volumes, GetVolumeID from the VMS will return the volume id by looking up for volume path from cache